### PR TITLE
chore!: make minio client optional and improve initialization

### LINKED ISSUE #429: Optimize Minio Client Usage by Reusing Across Requests
## Description

Currently, our application creates a new Minio client (`miniogo.Client`) for each request. This approach can lead to performance inefficiencies due to the overhead associated with repeatedly initializing new clients, especially under high load.

## Proposed Change

Refactor the application to instantiate a single Minio client and reuse it across multiple requests. This change aims to reduce initialization overhead and improve overall performance by leveraging connection pooling and reducing setup times.

## Benefits

- **Reduced Latency:** Minimizes the time spent in client setup for each request.
- **Improved Resource Utilization:** Lessens the burden on system resources by reducing the number of client instances.
- **Enhanced Performance:** Potentially increases throughput by reusing existing HTTP connections.

## Implementation Outline

1. **Create a `MinioService` struct** that includes a `miniogo.Client` as a field.
2. **Instantiate `MinioService`** at application startup with a single Minio client.
3. **Refactor existing code** to use the Minio client from `MinioService` instead of creating new instances.

## Code Snippet

```go
type MinioService struct {
    Client miniogo.Client
}
func NewMinioService(endpoint, accessKey, secretKey string) (MinioService, error) {
    cli, err := miniogo.New(endpoint, &miniogo.Options{
        Creds: credentials.NewStaticV4(accessKey, secretKey, ""),
        Secure: false,
})
if err != nil {
   return nil, err
}
return &MinioService{
       Client: cli,
    }, nil
}
```

### PR DESCRIPTION
**Pull request by mojtaba-esk**
Closes #429 

## Summary by CodeRabbit

- **New Features**
  - Introduced a consistent Minio client initialization to improve performance and reliability.

- **Refactor**
  - Renamed fields in various modules for better clarity (e.g., `K8s` to `K8sClient`, `Minio` to `MinioClient`).
  - Added a validation step in initialization functions to ensure proper configuration.

- **Bug Fixes**
  - Improved error handling with new error variables to address missing or mismatched configurations.

- **Tests**
  - Updated test cases to reflect changes in Minio and K8s client initialization.
  - Added new test scenarios for enhanced coverage.

### CODE REVIEW

File: pkg/minio/minio.go
Comment (MSevey) [inline pkg/minio/minio.go:47]: I'd call this `client`. I think `cli` is confusing since that is most often referred to for command line interface

File: pkg/minio/minio.go
Comment (MSevey) [inline pkg/minio/minio.go:187]: If we are refactoring this, why not refactor to actually have a `New` method?  Instead of trying to, or checking for initialization on every call of this method?

**Review by MSevey — CHANGES_REQUESTED** [review]
We should update this to have a `New` or `Init` method that is called once as opposed to requiring `init` to be called by every method.

File: pkg/minio/minio.go
Comment (mojtaba-esk) [reply pkg/minio/minio.go:187]: that makes sense, we can do the check in the knuu pkg and if user wants to use it, then we deploy it and start to use it

Comment (MSevey) [reply pkg/minio/minio.go:192]: Will you make these updates then? That was the original scope of the issue you made anyways. 
Or do you want to do that in a follow up?

Comment (mojtaba-esk) [reply pkg/minio/minio.go:192]: I will do them in this PR.

File: pkg/minio/minio.go
Comment (mojtaba-esk) [reply pkg/minio/minio.go]: done

File: pkg/minio/minio.go
Comment (mojtaba-esk) [reply pkg/minio/minio.go]: It is done, could you please have a look?

File: pkg/minio/minio.go
Comment (MSevey) [inline pkg/minio/minio.go:60]: leftover println

Comment (mojtaba-esk) [reply pkg/minio/minio.go]: fixed it

File: e2e/tshark/tshark_test.go
Comment (MSevey) [inline e2e/tshark/tshark_test.go:52]: you shouldn't need this anymore right? because you passed in the `MinioEnabled: true` option. 
This would overwrite the minio client that was already created. 

If overwriting was the goal, we should add a comment explaining why we are doing that.

Comment (mojtaba-esk) [reply e2e/tshark/tshark_test.go]: fixed

File: pkg/knuu/knuu.go
Comment (MSevey) [inline pkg/knuu/knuu.go:45]: I don't think we need the boolean. We can just have the `Minio` instance. 
Since it isn't a required service, developers can pass in `nil`. 
If they plan to use it, they just initialize the minio client and pass that in with the options.

File: pkg/knuu/knuu.go
Comment (MSevey) [inline pkg/knuu/knuu.go:75]: Build on comment, this would get removed.

File: pkg/knuu/knuu.go
Comment (MSevey) [inline pkg/knuu/knuu.go:59]: Building on comment, we would revert this.

File: pkg/knuu/knuu.go
Comment (MSevey) [inline pkg/knuu/knuu.go:201]: Building on comment, we would still remove this in order to make defining the `MinioClient` optional.

File: pkg/knuu/knuu.go
Comment (MSevey) [inline pkg/knuu/knuu.go:228]: Building on comment, we wouldn't need this. This would become part of test helpers or something.

File: pkg/knuu/knuu.go
Comment (mojtaba-esk) [reply pkg/knuu/knuu.go:45]: Ideally we could do it, then we need to also ask user to pass k8sClient as well since we need it for deploying minio.
Also we might also question the enabling proxy which is also a boolean.
So wdyt?

Comment (mojtaba-esk) [reply pkg/knuu/knuu.go:45]: I think also receiving k8s client from user makes more sense

Comment (MSevey) [reply pkg/knuu/knuu.go:45]: The k8 client is already passed in by the user, so the user has that to initialize the minio client. 

It also causes less of a diff, since there isn't a new option. It is just the change in default behavior that is impacted. 

For how the `New` method is currently structured, I'd recommend accepting the actual clients vs flags. I wouldn't change the `ProxyEnabled` flag just yet though. We should review the New method again after this PR is merged and see if there are any other updates we want to make as part of the breaking release.

Comment (mojtaba-esk) [reply pkg/knuu/knuu.go]: Ok pushed some changes. k8sClient was accepted as an option and if it was not passed, a default client would have been initiated. Now, user has to pass it as well. I think now it is in a pretty good shape. thanks for pointing it out.
Agreed that the proxy is something more embedded in knuu and refactoring its option requires more thought.

File: pkg/knuu/knuu.go
Comment (MSevey) [inline pkg/knuu/knuu.go:50]: why did you pull this out of the options? we should just leave it in the options.

File: pkg/knuu/knuu.go
Comment (MSevey) [inline pkg/knuu/knuu.go:194]: this should stay in.

File: pkg/knuu/knuu.go
Comment (MSevey) [inline pkg/knuu/knuu.go:88]: We shouldn't put test helpers in production code. 
Test helpers can be put in the `e2e` package.

File: e2e/bittwister/suite_setup_test.go
Comment (MSevey) [inline e2e/bittwister/suite_setup_test.go:22]: if we revert your change of adding the k8 client as a top level argument and removing it from the `setdefaults` function, then you could revert this.

File: e2e/system/build_from_git_test.go
Comment (MSevey) [inline e2e/system/build_from_git_test.go:26]: same here, as to why we shouldn't have the k8client as a standalone argument.

File: pkg/knuu/knuu.go
Comment (mojtaba-esk) [reply pkg/knuu/knuu.go:50]: because this is mandatory to have and it is not an optional parameter.
If we want to have Minio and probably some other services in future initiated by the user, which is a good practice, then we need to  use the same `k8sClient` for both. For example if we use one `k8sClient` for Minio and `Knuu` also makes its own, we will have two `k8sClinets` which introduces inconsistency also it might point to a different namespaces. So we have to extract it and ask user to pass it in.

File: pkg/knuu/knuu.go
Comment (mojtaba-esk) [reply pkg/knuu/knuu.go:88]: agreed

**Review by MSevey — CHANGES_REQUESTED** [review]
final clean up

File: e2e/system/build_from_git_test.go
Comment (mojtaba-esk) [reply e2e/system/build_from_git_test.go:27]: that's true. ill clean it up asap

File: pkg/knuu/knuu.go
Comment (MSevey) [inline pkg/knuu/knuu.go:197]: You're missing the default base case of setting TestScope when both are empty. 
```suggestion
	if k.TestScope == "" && k.K8sClient != nil {
		k.TestScope = k.K8sClient.Namespace()
	} else if k.TestScope == "" {
		t := time.Now()
		k.TestScope = fmt.Sprintf("%s-%03d", t.Format("20060102-150405"), t.Nanosecond()/1e6)
```

File: pkg/knuu/knuu_test.go
Comment (MSevey) [inline pkg/knuu/knuu_test.go:69]: per my original comment, this case can now be reverted as `TestScope` shouldn't be required.

File: pkg/knuu/knuu_test.go
Comment (MSevey) [inline pkg/knuu/knuu_test.go:137]: this should then be removed. as this is the base case

File: pkg/knuu/knuu_test.go
Comment (MSevey) [inline pkg/knuu/knuu_test.go:198]: this should be default empty options.

File: pkg/knuu/knuu_test.go
Comment (MSevey) [inline pkg/knuu/knuu_test.go:192]: this can be removed as this is just the base case that should be valid.

File: pkg/knuu/knuu.go
Comment (MSevey) [inline pkg/knuu/knuu.go:176]: This isn't actually a problem. Now that you've moved the validation logic out you can see that this is simply the base case. You should be able to not set these. 

The case you want to check is this:
```suggestion
	if opts.TestScope != "" && opts.K8sClient != nil && opts.TestScope != opts.K8sClient.Namespace() {
		return ErrTestScopeMistMatch
	}
```

File: e2e/system/build_from_git_test.go
Comment (MSevey) [inline e2e/system/build_from_git_test.go:26]: fixing the validation logic will revert this.

File: e2e/system/build_from_git_test.go
Comment (MSevey) [inline e2e/system/build_from_git_test.go:67]: Why did we need to add this test to this PR?

File: pkg/knuu/knuu_test.go
Comment (MSevey) [inline pkg/knuu/knuu_test.go:82]: To help catch some of the logic errors we are uncovering in this PR, I'd recommend you update this test struct to accept the actual error, not just whether you expect an error.

File: pkg/knuu/knuu_test.go
Comment (MSevey) [inline pkg/knuu/knuu_test.go:169]: this should be `err` and then we set the actual error we expect. This helps enough we are actually hitting the expected error and not just any error.

File: pkg/knuu/knuu_test.go
Comment (MSevey) [reply pkg/knuu/knuu_test.go:192]: the test case should be validating that when they are both set but don't match, it errors.

File: pkg/knuu/knuu.go
Comment (mojtaba-esk) [reply pkg/knuu/knuu.go]: well it was like this before, but in one of the comments you told me to move this default test scope to the e2e pkg: 
https://github.com/celestiaorg/knuu/pull/434/files/41fff2e769f1e40f1a336d265a1579265647fc47#diff-04a1accd1c12dafffe8d3ecc43a623cd0bf22d03d3561f7e04e296e64962048eR8-R11
that's the reason for receiving it from the user. I am more fan of making it easier for the users, so will put it back.

File: pkg/knuu/knuu.go
Comment (mojtaba-esk) [reply pkg/knuu/knuu.go:197]: well it was like this before, but in one of the comments you told me to move this default test scope to the e2e pkg: 
https://github.com/celestiaorg/knuu/pull/434/files/41fff2e769f1e40f1a336d265a1579265647fc47#diff-04a1accd1c12dafffe8d3ecc43a623cd0bf22d03d3561f7e04e296e64962048eR8-R11
that's the reason for receiving it from the user. I am more fan of making it easier for the users, so will put it back.

File: e2e/system/build_from_git_test.go
Comment (mojtaba-esk) [reply e2e/system/build_from_git_test.go:67]: because we refactored how user enables Minio and Minio is required by the builder when builds on directory context and since this test modifies the built image, it does it in a tmp directory which is local.

File: pkg/knuu/knuu.go
Comment (MSevey) [reply pkg/knuu/knuu.go:197]: Sorry, the name of this variable and the function is confusing. 
When it was refactored it looked like a testing variable, which is why I made the original comment. 
Knuu is a Test framework, which makes sense why this is called TestScope. I see that now. 

Maybe in a follow up we could rename this to just be Scope to avoid future confusion

File: e2e/system/build_from_git_test.go
Comment (MSevey) [reply e2e/system/build_from_git_test.go:67]: ah ok, thanks.

File: pkg/knuu/knuu.go
Comment (mojtaba-esk) [reply pkg/knuu/knuu.go:197]: yeah that's a good idea to rename it.

File: e2e/system/build_from_git_test.go
Comment (mojtaba-esk) [reply e2e/system/build_from_git_test.go:28]: it is done in `knuu.New`

**Review by MSevey — APPROVED** [review]
Nice!

**Pull request merged**
